### PR TITLE
Add a new thin abstraction layer over the creation of resolver instances

### PR DESCRIPTION
**Pull request by laeubi**
Currently Equinox directly reference the implementation of the (felix) resolver even though it never uses any specific API (except the logger), this makes it currently hard to allow other implementations especially as the is an unfortunate relation between the startup of launches and tests, so if another  implementation is used and might be buggy the framework possibly not start up and tests can not be executed or debugged.

This now adds a thin abstraction layer over the creation of resolver instances that basically distinguish the three cases we have where an instance is created allowing to later e.g. return dynamically based on the context (e.g. testrun) a different resolver instance.

### Background

Of the past months I struggled with analyze and optimize the current resolver implementation that works fast most of the time but in some cases can badly fail, see for example:

- https://github.com/eclipse-equinox/equinox/pull/1148
- https://github.com/eclipse-equinox/equinox/pull/1125
- https://github.com/eclipse-equinox/equinox/pull/469

I made several attempts already (and many failed) to improve this and sometimes it is possible to improve for a certain case but often it reveals then other corner cases and I identified some blockers that hinder making progress:

1. There is a high inherited risk to break everything, doing it "wrong" not only break tests it sometimes prevents the whole SDK and test launches from startup, this makes it hard to develop and hard to merge incremental improvements
2. Due to the Felix Impl is (intentionally) a separated place some useful methods in Equinox can only be indirectly called.
3. Some of the problems seem inherit by the used datastructure but major refactorings are hard to accomplish
4. Some limitations are imposed by the OSGi specification that do not matter for internal use cases.

To mitigate I would like to get this abstraction in place what will allow:

- Let dynamically exchange the resolver at runtime, so it is possible to run tests multiple times with different resolver implementations
- We can create different resolver instances per usecase (and maybe even optimized for a certain use-case!), e.g. think about a resolver that is not capable of resolve dynamic or violates some contracts of what the Resolver service imposes
- It would be possible to ship with a "classic" and a "new" resolver implementation so there is less of a risk for releases as users can choose or use a fallback.

### CODE REVIEW

File: bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/ResolverFactory.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	 * Creates a resolver instance that is used in the Equinox framework directly,
+	 * this method is called whenever a resolve of bundles is performed
+	 * 
+	 * @param container the container this instance is created for to maybe further
+	 *                  customize the instance
+	 * @param logger    a logger to emit log messages
+	 * @param executor  an executor to use to perform session related tasks
+	 * @return a new resolver instance
+	 */
+	public static Resolver createFrameworkResolver(ModuleContainerAdaptor container, Logger logger, Executor executor) {
```
[the remark this answers is not in the document]
Comment (laeubi) [reply bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/ResolverFactory.java]: The static methods are needed because wen need to perform this even before the framework starts (e.g. you need a resolved state to register services). So my (intermediate) plan would be to use the properties of the container to specify what class should be loaded so I can give it when the container starts (similar to system-packages).

Regarding the logger, that would generally be possible / useful I just don't want to make it to complicated here.

[the remark this answers is not in the document]
Comment (laeubi) [reply bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/ResolverFactory.java]: I first though about that, but it makes the code a lot more work as it should, e.g. we need to define an interface, then load all implementations and then choose if there are many what to use. 

So for now I want it really stupid simple and maybe we later make it more flexible e.g. to allow open it up to other (non internal) use-cases where service loader seems quite a good approach.

[the remark this answers is not in the document]
Comment (tjwatson) [reply bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/ResolverFactory.java]: > One thing that currently bugs me a bit that we even have EquinoxContainer + ModuleContainerAdaptor in the sigature

Right, why do you pass the `ModuleContainerAdaptor` here?  It doesn't seem to be used.
